Specify precise file dependencies in `Makefile`

jakelishman opened the pull request:
This refactors the C components of the Makefile significantly to better isolate three types of logic:

- what files actually need to be built and output
- how to build each file
- where each file should be "installed" for the output distribution

This is motivated because the `Makefile` was already missing several include files and directories that needed to be copied over or created, and we intend to make further changes to the C library distribution logic in the future.

When the `include/qiskit/` distribution subdirectory was added, the rules that construct it and the contained files were appended to existing rules for `qiskit.h`, which meant that make did not understand the true dependency chain, and so did not reliably consider the sources dirty because it didn't know about them. This meant that it might not create necessary intermediate states, and might not update changed files in a pre-built directory.

We also remove the top-level `CMakeLists.txt` file because the top-level of the repository doesn't actually use CMake at all; we previously only used it to feed in the otherwise out-of-tree locations of the Qiskit library and include files.  This moves all the CMake logic into `test/c` as an isolated project, and has the Makefile (which is what actually arranges for the files to exist in the first place) fill in the search locations for them.


### Summary



### Details and comments

I'm not an expert in `make`, so there might be simpler ways to do this, but I was deliberately avoiding globbing files in `crates/cext/include` (you have to manually specify which ones should get copied over).  The general principle is that _every_ direct dependency (file, directory, build process, etc) is explicitly listed in the dependencies for any given recipe, even if there is no build step (like for `crates/cext/include/complex.h`), so `make` knows how to tell if something is dirty.

The motivation for this PR is that I had a pre-built structure from before #14340, which had `dist/c/include/qiskit.h` but not the directory `dist/c/include/qiskit/`.  Since `dist/c/include` already existed, and the `dist/c/include/qiskit/` directory was only built if its parent was necessary, the `Makefile` failed to copy the outputs.

I used the LLM below to help remind me of GNU Make options and built-in variables.

AI disclosure: IBM Bob 0.0.14 with claude-3-5-sonnet-20241022

jakelishman commented on Makefile:
```diff
@@ -124,21 +135,25 @@ $(C_DIR_LIB):
+ctest: $(C_DIR_INCLUDE)/qiskit
 	cargo rustc --crate-type cdylib -p qiskit-cext
 	cp target/qiskit.h $(C_DIR_INCLUDE)/qiskit.h
 	cp crates/cext/include/complex.h $(C_DIR_INCLUDE)/qiskit/complex.h
```
This isn't correct (I'm not sure why I made the one change I did), but right now the `Makefile` is not organised well enough to conditionally choose whether to include the `--release` flag or not, so I've left sorting that out for a follow-up.

Cryoris reviewed (commented):
If I'm using this locally running either of `make c` or `make cheader` always rebuilds and copies both the headers and the library (even when requesting the header only). Is this intentional?

jakelishman commented:
Yes, because the header is dependent on the build of the library.  The library is always rebuilt if it's in a requirement because `cargo` manages the dirty/clean status of it, and `make` doesn't have enough information to tell.

jakelishman commented on Makefile:
```diff
@@ -124,21 +135,25 @@ $(C_DIR_LIB):
-	cp target/qiskit.h $(C_DIR_INCLUDE)/qiskit.h
-	cp crates/cext/include/complex.h $(C_DIR_INCLUDE)/qiskit/complex.h
+target/qiskit.h: $(C_LIBQISKIT)
+$(C_QISKIT_H): $(C_DIR_INCLUDE) $(C_LIB_CARGO_PATH) target/qiskit.h
```
Julien: explicitly, it's these two lines you're asking about.  Line 144 says "the file `target/qiskit.h` is generated during the build of the `libqiskit` file" and line 145 says "`/dist/c/include/qiskit.h` depends on `/dist/c/include` existing, the cargo library path existing [I'm not sure why? - that was pre-existing and maybe unnecessary], and `target/qiskit.h` being fresh".

jakelishman committed:
Specify precise file dependencies in `Makefile`

This refactors the C components of the Makefile significantly to better
isolate three types of logic:

- what files actually need to be built and output
- how to build each file
- where each file should be "installed" for the output distribution

This is motivated because the `Makefile` was already missing several
include files and directories that needed to be copied over or created,
and we intend to make further changes to the C library distribution
logic in the future.

When the `include/qiskit/` distribution subdirectory was added, the rules
that construct it and the contained files were appended to existing
rules for `qiskit.h`, which meant that make did not understand the true
dependency chain, and so did not reliably consider the sources dirty
because it didn't know about them. This meant that it might not create
necessary intermediate states, and might not update changed files in a
pre-built directory.

We also remove the top-level `CMakeLists.txt` file because the top-level
of the repository doesn't actually use CMake at all; we previously only
used it to feed in the otherwise out-of-tree locations of the Qiskit
library and include files.  This moves all the CMake logic into `test/c`
as an isolated project, and has the Makefile (which is what actually
arranges for the files to exist in the first place) fill in the search
locations for them.

I used the LLM below to help remind me of GNU Make options and built-in
variables.

AI disclosure: IBM Bob 0.0.14 with claude-3-5-sonnet-20241022

jakelishman commented:
I've just come back to this PR and fairly significantly reworked it, in part because I had new ideas since I first opened it, and we've had some changes in the build system since then.

I want to make some more major changes to the C build system, and this new layout of the `Makefile` will significantly help me with that.

Also, I asked Bob for a code review, and it told me:

> Recommendation: APPROVE ✅
>
> This is a high-quality refactoring that significantly improves maintainability and clarity of the build system. The changes are well-thought-out, properly documented, and follow best practices. The breaking changes are justified by the architectural improvements.

so the robots think I'm doing a good job lol

It should be mentioned that breaking changes in general _are not_ justified by architectural improvements - users don't care if we made life easier for ourselves, they care if we made life better for them.  But in this case, Bob's pretty much wrong about it being a breaking change anyway - the top-level `CMakeLists.txt` file wasn't really usable for anything much, and `make ctest` was the documented correct way to run the tests, which still works.

jakelishman committed:
Move C coverage logic into `Makefile`

The logic of running the C test suite with coverage enabled was
duplicated between the `Makefile` and the Coverage runner, making it
harder to keep in sync.  This makes sure they'll be unified.

jakelishman committed:
Correct newly non-nested output directory

mrossinek commented:
_I've request a review from myself to make sure I get back to this and learn from it to improve the `qiskit-fermions` setup wherever possible, but a pending review from me should by no means be blocking_ :+1:

jakelishman commented:
Max: I'm actually expecting to use this refactor as a stepping stone before changing the build system a bit more completely.

The use of Make here is a little awkward - Make is typically best at things like C, where most of the worker targets are actually just a list of files that are each built directly and individually by one command invocation (like a C compiler or a linker).  Here, we're using Make basically just as a command runner to orchestrate _different_ build systems (`cargo` and `cmake`), so we're getting stuck with the tricky Make syntax and logic, and not actually benefiting from its main strengths (calculating the minimal set of recipes that need to be refreshed and parallelising them).  Something like [`invoke`](https://www.pyinvoke.org/) or [`just`](https://just.systems/man/en/) might be more appropriate, but at the downside of them being another tool everybody would have to install.

(I'm not intending to replace Make btw, I just mean that I'm going to change a bunch of the recipes.)

eliarbel commented on CONTRIBUTING.md:
```diff
@@ -714,8 +714,10 @@ well.
+The C API test suite is located at `test/c/`.  This is a CMake project and uses
+CMake's `ctest` runner.  To build and run the tests, use the `ctest` recipe in
+the top-level `Makefile`, which you can run with
+
```
Not directly related to this PR, but since we're at it: In my local Linux environment, I need to use `LD_LIBRARY_PATH` to point to the Python shared object for the tests to run. I guess this is also the case on other platforms. Until we make the C API fully Python independent, should we add a comment about this here?

jakelishman replied:
Oh, I didn't know that to be a problem, and we don't do anything like that in CI to get it to work (nor on my machine, but I've got a Mac).  I'm not sure what's going on with your environment to require that.

mrossinek replied:
FWIW I also don't need that for the pure `ctest` target on Fedora over here

eliarbel replied:
Thanks Max. Yes, I also got a validation from Gadi this morning that he doesn't need this as well. This is something unique in my env.  Thanks for validating.

eliarbel commented on Makefile:
```diff
@@ -101,60 +126,90 @@ else
+cheader: $(C_INCLUDE_FILES_OUT_GENERATED) $(C_INCLUDE_FILES_OUT_MANUAL)
+.PHONY: clib
+clib: $(C_LIBQISKIT_OUT)
+.PHONY: c
```
What do you think about adding a `c-dev` rule (that is one that uses `build-clib-dev`)? This would be more developer friendly when working on e.g. a standalone c file.  The builds would be faster, and developers would also be able to use the same env variables and compiler flags against the distribution directory, instead of pointing to Cargo's target artifacts.

jakelishman replied:
I can do, though I worry it'll interact a little weirdly with the Make "freshness" rules, because the same file (`dist/c/lib/libqiskit.so`) will have more than one way to generate it.  So if you run `make clib-dev` followed by `make clib`, then the `make clib` wouldn't actually regenerate the file into release mode, because it'd appear to be fresh.  You be able to force it with a `make cclean` in the middle, but that's a trip hazard.

I can add it if you think it's valuable with those caveats, though.

eliarbel replied:
That's right, both flows will use the same identical distro structure, but I think for a developer that's still a small risk compared to the benefit of having the same distro output but with much shorter compilation cycle times. 
So yes, I think it would be valuable to have.

jakelishman replied:
Ok done in b886897647.

eliarbel commented on Makefile:
```diff
@@ -101,60 +126,90 @@ else
 	cmake --build $(C_DIR_TEST_BUILD)
 	# -V ensures we always produce a logging output to indicate the subtests
 	# -C Debug is needed for windows to work, if you don't specify Debug (or
 	#  release) explicitly ctest doesn't run on windows
```
```suggestion
    # `-S` specifies the source (including the `CMakeLists.txt` file, `-B` is
    # where to put the build files, including the generated CMake stuff.  See
    # the `CMakeLists.txt` file for the build variables.
	cmake -S$(C_DIR_TEST) -B$(C_DIR_TEST_BUILD) \
		-DCARGO_LIB_DIR=$(abspath $(C_DIR_CARGO_TARGET))/debug \
		-DQISKIT_INCLUDE_PATH=$(abspath $(C_DIR_OUT_INCLUDE))
    # Actually build the test.
	cmake --build $(C_DIR_TEST_BUILD)
    # -V ensures we always produce a logging output to indicate the subtests
    # -C Debug is needed for windows to work, if you don't specify Debug (or
    #  release) explicitly ctest doesn't run on windows
```
Just to make these comments Make comments instead of Shell comments.

jakelishman replied:
Done in a461232f2 (though I got rid of the indentation entirely, because mixing tabs and spaces for the same indentation level is heresy lol)

eliarbel commented on Makefile:
```diff
@@ -101,60 +126,90 @@ else
-	rm -rf $(C_DIR_OUT) $(C_DIR_TEST_BUILD)
-	rm -f target/qiskit.h
-	cargo clean
+	rm -rf $(C_DIR_OUT) $(C_DIR_TEST_BUILD) $(C_INCLUDE_FILES_ABS_GENERATED)
+	cargo clean --package qiskit-cext
```
I'm wondering why we are deleting the Cargo artifacts here and not just the Makefile ones? Is there a justification for this from the C API build/test/coverage perspectives? I'm thinking about a workflow where I develop the C API, and after I'm done I want to clean the dist directory but don't want the Cargo artifacts to also go away. (Yeah, this are just `qiskit.h` and the cext crate, but still I was wondering...)

jakelishman replied:
The Makefile is just acting as a command runner that calls `cargo`, so if we don't clean out the `cargo` artifacts, then `make cclean` followed by `make clib` (or whatever) wouldn't actually force a rebuild of `cext`, which is what (I'm fairly sure) Julien intended by the original rule.

eliarbel replied:
I thought that since `clib` and all its dependency rules are basically PHONY targets, leading to running `cargo` anyway each time `make clib` (or `make c`) is called, Cargo will know when a rebuild is needed. Isn't this the case?

jakelishman replied:
The full answer is: I don't _know_ why Julien wrote it this way originally, but I suspect the answer is something to do with the build script dropping dirty files in the `target` directory, and wanting to be 100% sure the complete rebuild would happen.

eliarbel commented on Makefile:
```diff
@@ -101,60 +126,90 @@ else
+.PHONY: clib
+clib: $(C_LIBQISKIT_OUT)
+.PHONY: c
+c: cheader clib
```
Just a thought: `qiskit.h` is generated twice here. 
What if the rule would be this instead: 
```makefile
c: clib cheader
```
Together with making the `qiskit.h` target like this: 
```makefile
$(C_INCLUDE_FILES_ABS_GENERATED): 
	build-cheader
```

jakelishman replied:
That's already the rule for `C_INCLUDE_FILES_ABS_GENERATED`, I think?  Well, except that in yours the dependency is misaligned as a recipe step rather than a dependency (so `make` would try and call `build-cheader` as a shell command, which would fail).  If you intended that to be `$(MAKE) build-cheader`, I don't think it'd make a difference though - either way, `build-cheader` and `clib` are both `PHONY`, so they're also marked out-of-date and Make will always try and re-run them.  We need that to be the case, because we can't actually spell the file-level dependencies to `make` - that's `cargo`'s job, so we just tell `make` to consider it always dirty and let `cargo` actually check.

I don't think either your way or my way currently can avoid generating the header file twice unless we duplicate a bunch of internal logic right now.  But that said, the reason I came back to sorting out this `Makefile` is that I'm about to shift the header-file generation logic out of the `cext` build script, which will naturally make it possible to only generate the header file once (although we're still unlikely to be able to avoid having a `.PHONY` target internally because only `cargo` would be able to tell if the `cext` sources are actually dirty).

jakelishman replied:
Generating the include files is rather inexpensive, though, so it shouldn't be too painful.

eliarbel replied:
I think in my suggestion the header file should be generated once and then copied with the `$(C_INCLUDE_FILES_OUT_GENERATED)` rule.  But never mind, this is really minor and you're going to change this part anyway, so not worth dwelling on it.  

> If you intended that to be $(MAKE) build-cheader

Oh sure, my bad

jakelishman replied:
Ah yeah, I see what you mean - the submake "hides" the phony dependency. Yeah, I think you're right it'd work.

If it's ok I'll just leave it for now, though - once I'm back home later today I'll have a PR that obsoletes this concern anyway.

eliarbel reviewed (commented):
This looks very nice, Jake. I like how the variables are now more modularly organized and the extra documentation. But most importantly 😉: kudos for making the `cheader` rule not triggering a release build every time 🏆. (it was annoying when rendering docs locally; now I don't have to remove the `--release` flag in the Makefile just for this)

Overall this looks good and a good step forward. I left a few minor comments and suggestions inside. 

One last thing to ask: why did you choose to list the header files in `qiskit/include` explicitly in `C_INCLUDE_FILES_MANUAL` rather than collecting them using e.g. a wildcard based-rule? Would that be better for reducing future maintenance toll?

jakelishman committed:
Avoid comments within recipes

jakelishman commented:
> why did you choose to list the header files in `qiskit/include` explicitly in `C_INCLUDE_FILES_MANUAL` rather than collecting them using e.g. a wildcard based-rule? Would that be better for reducing future maintenance toll?

Mostly because I don't hugely like wildcards, especially in build systems, and prefer to be painfully explicit.  In this case, omission of a file should fail safe, because the `make ctest` rule should fail to compile.

Wildcards only match the current directory in GNU Make, and there's no (afaik) equivalent of bash's `**` with `shopt -s globstar`, so we'd still need to list out the directory structure of `include` manually, and I thought it wasn't too far a stretch beyond that just to write out the files too.  I don't feel hugely strongly, though.

jakelishman committed:
Add `clib-dev` recipe

This allows building and installing the library in debug mode into the
same file path as `clib` uses.

mrossinek commented on test/c/CMakeLists.txt:
```diff
@@ -3,6 +3,25 @@
 # https://cmake.org/cmake/help/book/mastering-cmake/chapter/Testing%20With%20CMake%20and%20CTest.html
 cmake_minimum_required (VERSION 3.27)
```
Random comment: in https://github.com/Qiskit/qiskit-fermions/pull/15 I'm downgrading the minimum required version to 3.22 without any problems. I've had to do this manually on various servers in the past since their shipped cmake was not new enough.

Looking at [repology](https://repology.org/project/cmake/versions) shows that at least Ubuntu 22.04 (which is reaching EOL in June 2027) still ships with cmake 3.22. I'm not sure about RHEL 8 but would expect something similar there, surely not 3.27+.

jakelishman replied:
I'm happy to lower this if we can get away with it. I can't imagine we use many overly sophisticated CMake features.

Fwiw, `cmake` is packaged on PyPI in wheels, so it you need it, you can set a build dependency on it to get a recent version.

mrossinek reviewed (commented):
Some tiny comments.. This looks great! :+1:

eliarbel reviewed (approved):
Thanks Jake, this looks good and ready to merge. If at some point we'll want to lower the required CMake version we can handle it in a separate PR.
